### PR TITLE
Add desktop serialize test case

### PR DESCRIPTION
**Pull request by JimboWei**
@shmao Please help review the code.

### CODE REVIEW

**Comment by JimboWei** [comment]
@shmao some builds failed because there are test case using:

1.  `string Data = new DateTime().ToLongDateString();` got different results 
Monday, January **1**, 0001  (baseline)
Monday, January **01**, 0001
2. `DateTime().ToShortDateString()` got different results
1/1/**0001**  (baseline)
1/1/**01**
Do you have any idea ?

**Comment by JimboWei** [comment]
@shmao I have fixed the test case by changing the dateTime string to normal string content, but there is still a case named `DCS_BasicRoundtripDCRDefaultCollections` failed. Can you help to take a look ?

Error  message :

1. expected: <_arrayList z:Id=\"2\" **z:Size=\"1\"** xmlns:a=\"http://schemas.microsoft.com/2003/10/Serialization/Arrays\">
actual: <_arrayList z:Id=\"2\" xmlns:a=\"http://schemas.microsoft.com/2003/10/Serialization/Arrays\">

2. expected: <_hashtable z:Id=\"6\" **z:Size=\"1\"** xmlns:a=\"http://schemas.microsoft.com/2003/10/Serialization/Arrays\">
actual: <_hashtable z:Id=\"6\" xmlns:a=\"http://schemas.microsoft.com/2003/10/Serialization/Arrays\">

**Comment by shmao** [comment]
@dotnet-bot test this please

**Comment by shmao** [comment]
@JimboWei can you please open an issue tracking DCS_BasicRoundtripDCRDefaultCollections failing in ReflectionOnly? Let's skip the test using '#if !ReflectionOnly` for now.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 1]
```diff
@@ -3055,6 +3054,2031 @@ public static void DCS_BasicRoundtripDCRDefaultCollections()
...
+    [Fact]
+    public static void DCS_BasicPerSerializerRoundTripAndCompare_IObjectRef()
+    {
+        var setting = new DataContractSerializerSettings()
+        {
+            DataContractResolver = new SerializationTestTypes.SimpleResolver()
+        };
+
+        //DCExplicitInterfaceIObjRef
+        string baseline1 = @"<ObjectContainer xmlns=""http://schemas.datacontract.org/2004/07/SerializationTestTypes"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance""><_data i:type=""a:SerializationTestTypes.DCExplicitInterfaceIObjRef***"" xmlns:a=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.DCExplicitInterfaceIObjRef***""><data z:Id=""i1"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/""><Data z:Ref=""i1""/></data></_data><_data2 i:type=""a:SerializationTestTypes.DCExplicitInterfaceIObjRef***"" xmlns:a=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.DCExplicitInterfaceIObjRef***""><data z:Ref=""i1"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/""/></_data2></ObjectContainer>";
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: These test cases are very similar. It'd be better to refactor the pattern into a method, e.g.

```c#
        private static void TestObjectInObjectContainerWithSimpleResolver<T>(T o, string baseline) 
        {
            var setting = new DataContractSerializerSettings()
            {
                DataContractResolver = new SerializationTestTypes.SimpleResolver()
            };

            var value = new SerializationTestTypes.ObjectContainer(o);
            var actual = SerializeAndDeserialize(value, baseline, setting);
            SerializationTestTypes.ComparisonHelper.CompareRecursively(value, actual);
        }
```
And we can call the method like,

```c#
TestObjectInObjectContainerWithSimpleResolver(new SerializationTestTypes.DCExplicitInterfaceIObjRef(true),
    "baseline string....");
```

**Comment by JimboWei** [comment]
I have open an issues #21975 about failed case named DCS_BasicRoundtripDCRDefaultCollections.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs (hunk 1 above)
Comment (JimboWei) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: I have refactored the code as you suggested.

**Comment by shmao** [comment]
@JimboWei I have a PR, https://github.com/dotnet/corefx/pull/22286, for fixing the issue with ReflectionOnly mode. After that, all the tests in `DCS_BasicPerSerializerRoundTripAndCompare_SampleTypes_FailInReflectionOnly` should pass.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 2]
```diff
@@ -3050,6 +3050,748 @@ public static void DCS_BasicRoundtripDCRDefaultCollections()
...
+
+        TestObjectInObjectContainerWithSimpleResolver(new SerializationTestTypes.SampleICollectionExplicitWithDC(true), @"<ObjectContainer xmlns=""http://schemas.datacontract.org/2004/07/SerializationTestTypes"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance""><_data z:Id=""i1"" i:type=""a:SerializationTestTypes.SampleICollectionExplicitWithDC***"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/"" xmlns:a=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.SampleICollectionExplicitWithDC***""/><_data2 z:Ref=""i1"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/""/></ObjectContainer>");
+
+        TestObjectInObjectContainerWithSimpleResolver(new SerializationTestTypes.SampleICollectionExplicitWithoutDC(true), @"<ObjectContainer xmlns=""http://schemas.datacontract.org/2004/07/SerializationTestTypes"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance""><_data i:type=""a:SerializationTestT
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Can you please open an issue for this and modify this attribute?

Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: The tests would pass with https://github.com/dotnet/corefx/pull/22286. Can you please change the method name or move the tests into `DCS_BasicPerSerializerRoundTripAndCompare_SampleTypes`?

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 3]
```diff
@@ -3051,6 +3050,846 @@ public static void DCS_BasicRoundtripDCRDefaultCollections()
...
...rWithSimpleResolver(new SerializationTestTypes.CB1(true), @"<ObjectContainer xmlns=""http://schemas.datacontract.org/2004/07/SerializationTestTypes"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance""><_data z:Id=""i1"" i:type=""a:SerializationTestTypes.CB1***"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/"" xmlns:a=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.CB1***""><anyType z:Id=""i2"" i:type=""b:SerializationTestTypes.PublicDC***"" xmlns:b=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.PublicDC***""><Data>55cb1688-dec7-4106-a6d8-7e57590cb20a</Data></anyType><anyType z:Id=""i3"" i:type=""b:SerializationTestTypes.PublicDCClassPublicDM***"" xmlns:b=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.PublicDCClassPublicDM***""><Data i:nil=""true""/></anyType><anyType z:Id=""i4"" i:type=""b:SerializationTestTypes.PublicDCClassPrivateDM_DerivedDCClassPublic***"" xmlns:b=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.PublicDCClassPrivateDM_DerivedDCClassPublic***""><_data/></anyType></_data><_data2 z:Ref=""i1"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/""/></ObjectContainer>");
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Why some lines are commented out? Did they pass? Do we want to open issues for them?

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 4]
```diff
@@ -3051,6 +3050,846 @@ public static void DCS_BasicRoundtripDCRDefaultCollections()
...
+
+        TestObjectInObjectContainerWithSimpleResolver(new SerializationTestTypes.SampleICollectionExplicitWithCDCContainsPrivateDC(true), @"<ObjectContainer xmlns=""http://schemas.datacontract.org/2004/07/SerializationTestTypes"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance""><_data z:Id=""i1"" i:type=""a:SerializationTestTypes.SampleICollectionExplicitWithCDCContainsPrivateDC***"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/"" xmlns:a=""http://schemas.datacontract.org/2004/07/SerializationTestTypes.SampleICollectionExplicitWithCDCContainsPrivateDC***""><Item i:type=""b:dateTime"" xmlns=""Test"" xmlns:b=""http://www.w3.org/2001/XMLSchema"">0001-01-0
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Why are these in comments?

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs (hunk 3 above)
Comment (JimboWei) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: These lines can not pass when targeting uapaot, I have moved them into an new method and opened issues for them. I will remove the code.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs (hunk 4 above)
Comment (JimboWei) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: These lines can not pass when targeting uapaot, I have moved them into an new method and opened issues for them. I will remove the code.

**Comment by karelz** [comment]
Another older PR. What is the expected timeline to get it finished?

**Comment by JimboWei** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by shmao** [comment]
@JimboWei thanks for the PR. It LGTM. Can you please confirm if the tests passed on uap/uapaot run? Tests marked with [ActiveIssue] are fine.

**Comment by JimboWei** [comment]
@shmao I have confirmed the tests can pass both on uap and uapaot run.

**Comment by shmao** [comment]
Thanks @JimboWei . I'm merging the PR.

**Pull request merged**